sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcRelation.scala

```diff
@@ -0,0 +1,176 @@
+import org.apache.spark.Logging
+import org.apache.spark.annotation.DeveloperApi
+import org.apache.spark.mapred.SparkHadoopMapRedUtil
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.catalyst.expressions.Attribute
+import org.apache.spark.sql.hive.HiveMetastoreTypes
+import org.apache.spark.sql.types.StructType
+import org.apache.spark.sql.{Row, SQLContext}
+import org.apache.spark.sql.sources._
+import scala.collection.JavaConversions._
+
+
+private[sql] class DefaultSource extends FSBasedRelationProvider {
+
+  def createRelation(
+      sqlContext: SQLContext,
+      paths: Array[String],
+      schema: Option[StructType],
+      partitionColumns: Option[StructType],
+      parameters: Map[String, String]): FSBasedRelation ={
+    val partitionSpec = partitionColumns.map(PartitionSpec(_, Seq.empty[Partition]))
+    OrcRelation(paths, parameters,
+      schema, partitionSpec)(sqlContext)
+  }
+}
+
+
+private[sql] class OrcOutputWriter extends OutputWriter with SparkHadoopMapRedUtil {
+  var recordWriter: RecordWriter[NullWritable, Writable] = _
+  var taskAttemptContext: TaskAttemptContext = _
+  var serializer: OrcSerde = _
+  var wrappers: Array[Any => Any] = _
+  var created = false
+  var path: String = _
+  var dataSchema: StructType = _
+  var fieldOIs: Array[ObjectInspector] = _
+  var standardOI: StructObjectInspector = _
+
+
+  override def init(path: String,
+      dataSchema: StructType,
+      context: TaskAttemptContext): Unit = {
+    this.path = path
+    this.dataSchema = dataSchema
+    taskAttemptContext = context
+  }
+
+  // Avoid create empty file without schema attached
+  private def initWriter() = {
+    if (!created) {
+      created = true
+      val conf = taskAttemptContext.getConfiguration
+      val outputFormat = new OrcOutputFormat()
+      val taskId: TaskID = taskAttemptContext.getTaskAttemptID.getTaskID
+      val partition: Int = taskId.getId
+      val filename = s"part-r-${partition}-${System.currentTimeMillis}.orc"
+      val file = new Path(path, filename)
+      val fs = file.getFileSystem(conf)
+      val orcSchema = HiveMetastoreTypes.toMetastoreType(dataSchema)
+
+      serializer = new OrcSerde
+      val typeInfo: TypeInfo =
+        TypeInfoUtils.getTypeInfoFromTypeString(orcSchema)
+      standardOI = TypeInfoUtils
+        .getStandardJavaObjectInspectorFromTypeInfo(typeInfo)
+        .asInstanceOf[StructObjectInspector]
+      fieldOIs = standardOI
+        .getAllStructFieldRefs.map(_.getFieldObjectInspector).toArray
+      wrappers = fieldOIs.map(HadoopTypeConverter.wrappers)
+      recordWriter = {
+        outputFormat.getRecordWriter(fs,
+          conf.asInstanceOf[JobConf],
+          file.toUri.getPath, Reporter.NULL)
+          .asInstanceOf[org.apache.hadoop.mapred.RecordWriter[NullWritable, Writable]]
+      }
+    }
+  }
+  override def write(row: Row): Unit = {
+    initWriter()
+    var i = 0
+    val outputData = new Array[Any](fieldOIs.length)
```

Comment:
I guess we can reuse a mutable array here rather than create an array for each row object.